lib/CoreDisTools/coredistools.cpp

```diff
@@ -37,25 +37,98 @@
+  size_t BlockSize;
+  // The original base address of the code block.
+  uintptr_t Addr;
+
+private:
+  bool nearDiff(const BlockInfo &Block, const DiffControl &Control) const;
+  static bool diffError(const DiffControl &Control, const char *Mesg,
+                        const BlockIterator &Left, const BlockIterator &Right);
+};
+
+// A block iterator reoresents a code-point within a code block.
+// It represents an instruction within the block, and can move
+// forward to subsequent instructions via the advance() method.
+// The iterator can be in two modes:
+//  Not-Decoded: Before DecodeInstruction() is called at this
+//               code point
+//  Decoded: After the current instrction is Decoded. In this state,
+//           Inst is a valid MCInst and InstrSize is an actual non-zero
+//           length of the current instruction.
+class BlockIterator : public BlockInfo {
+public:
+  BlockIterator(const BlockInfo &Block)
+      : BlockInfo(Block), Inst(), InstrSize(0) {}
+  BlockIterator(const uint8_t *Pointer, size_t Size, uintptr_t Address = 0,
+                const char *BlockName = "")
+      : BlockInfo(Pointer, Size, Address, BlockName), Inst(), InstrSize(0) {}
+
+  void advance() {
+    assert(isDecoded() && "Cannot advance before Decode");
+    assert(InstrSize <= BlockSize && "Overflow");
+    Ptr += InstrSize;
+    Addr += InstrSize;
+    BlockSize -= InstrSize;
+
+    // Next instruction is not yet decoded
+    InstrSize = 0;
+  }
+
+  size_t isDecoded() const { return (InstrSize != 0); }
+  bool isBitwiseEqual(const BlockIterator &BIter) const;
+
+  // The machine instruction at this code point, after decode.
+  MCInst Inst;
+
+  // Why store the InstrSize separately, rather than obtaining
+  // it from Inst.Size()? This is because of a limitation in
+  // MCInst representation on certain architectures.
+  // Prefix instructions on X64 such as Lock prefix are
+  // encoded as an MCInst of zero size! When such a prefix is
+  // decoded, the actual decode InstrSize=1, but Inst.Size()=0
+  size_t InstrSize;
+};
+
 // Instruction-wise disassembler helper.
 // This utility is used to implement GcStress in CoreCLr
 // Adapted from LLVM-objdump
 
 struct CorDisasm {
 public:
+  CorDisasm(enum TargetArch Target) { TargetArch = Target; }
   bool init();
-  size_t disasmInstruction(size_t Address, const uint8_t *Bytes,
-                           size_t Maxlength, bool PrintAsm = false) const;
-  void printInstruction(const MCInst *MI, size_t Address, size_t InstSize,
-                        ArrayRef<uint8_t> Bytes) const;
-
-  CorDisasm(TargetArch Target) { TargetArch = Target; }
+  void decodeInstruction(BlockIterator &BIter) const;
+  size_t disasmInstruction(BlockIterator &BIter, bool PrintAsm = false) const;
+  void printInstruction(const BlockIterator &BIter) const;
 
 private:
   bool setTarget();
   bool verifyPrefixDecoding();
 
-  TargetArch TargetArch;
+  enum TargetArch TargetArch;
```

Comment:
OK, I'll rename the field as TheTargetArch -- so that it'll merge with #1030.
The rename would probably merge correctly even without this change.